File: usertests/kvm/host/src/device/bus.cpp
// SPDX-License-Identifier: MIT
// Portions of this file are adapted from KXemu's MMIO bus code.
// Original project: KXemu (MIT License), Copyright (c) 2024 HD-CSKX.
// Source repository: git@github.com:HD-CSKX/KXemu.git

#include "device/bus.hpp"

#include "kvm.hpp"
#include "linux_dtb.hpp"

#include <cstdio>
#include <limits>
#include <utility>

namespace kvm_host {
static bool checked_range_end(std::uintptr_t start, std::uintptr_t length, std::uintptr_t *end) {
    if (length == 0) {
        return false;
    }
    if (start > std::numeric_limits<std::uintptr_t>::max() - length) {
        return false;
    }

    *end = start + length;
    return true;
}

static bool ranges_overlap(std::uintptr_t left_start, std::uintptr_t left_length, std::uintptr_t right_start,
                           std::uintptr_t right_length) {
    std::uintptr_t left_end = 0;
    std::uintptr_t right_end = 0;
    if (!checked_range_end(left_start, left_length, &left_end) ||
        !checked_range_end(right_start, right_length, &right_end)) {
        return true;
    }

    return left_start < right_end && right_start < left_end;
}

bool Bus::map_area(const Kvm &kvm, std::uintptr_t guest_addr, std::uintptr_t length, std::uint8_t **mapped_host_addr) {
    std::uintptr_t end = 0;
    if (!checked_range_end(guest_addr, length, &end)) {
        std::fprintf(stderr, "bus map area has invalid range: addr=0x%lx length=0x%lx\n",
                     static_cast<unsigned long>(guest_addr), static_cast<unsigned long>(length));
        return false;
    }

    if (this->area_count_ == this->MAX_AREAS) {
        std::fprintf(stderr, "bus map area table is full\n");
        return false;
    }

    for (std::size_t i = 0; i < this->area_count_; i++) {
        if (ranges_overlap(guest_addr, length, this->areas_[i].guest_addr, this->areas_[i].length)) {
            std::fprintf(stderr, "bus map area overlaps existing range: new=[0x%lx,0x%lx) old=[0x%lx,0x%lx)\n",
                         static_cast<unsigned long>(guest_addr), static_cast<unsigned long>(end),
                         static_cast<unsigned long>(this->areas_[i].guest_addr),
                         static_cast<unsigned long>(this->areas_[i].guest_addr + this->areas_[i].length));
            return false;
        }
    }
    for (std::size_t i = 0; i < this->mmio_device_count_; i++) {
        if (ranges_overlap(guest_addr, length, this->mmio_regions_[i].guest_addr, this->mmio_regions_[i].length)) {
            std::fprintf(stderr, "bus map area overlaps mmio device: addr=0x%lx length=0x%lx\n",
                         static_cast<unsigned long>(guest_addr), static_cast<unsigned long>(length));
            return false;
        }
    }

    std::uint8_t *host_addr = nullptr;
    if (!kvm.map_area_raw(guest_addr, length, &host_addr)) {
        return false;
    }
    if (host_addr == nullptr) {
        std::fprintf(stderr, "ioctl(KVM_MAP_AREA) returned null host mapping\n");
        return false;
    }

    this->areas_[this->area_count_] = {guest_addr, length, host_addr};
    this->area_count_++;
    if (mapped_host_addr != nullptr) {
        *mapped_host_addr = host_addr;
    }
    return true;
}

bool Bus::add_mmio_device(std::uintptr_t guest_addr, std::uintptr_t length, std::shared_ptr<MmioDevice> device,
                          unsigned int id) {
    if (device == nullptr) {
        std::fprintf(stderr, "bus mmio device is null\n");
        return false;
    }

    std::uintptr_t end = 0;
    if (!checked_range_end(guest_addr, length, &end)) {
        std::fprintf(stderr, "bus mmio device has invalid range: addr=0x%lx length=0x%lx\n",
                     static_cast<unsigned long>(guest_addr), static_cast<unsigned long>(length));
        return false;
    }

    if (this->mmio_device_count_ == this->MAX_MMIO_DEVICES) {
        std::fprintf(stderr, "bus mmio device table is full\n");
        return false;
    }

    for (std::size_t i = 0; i < this->area_count_; i++) {
        if (ranges_overlap(guest_addr, length, this->areas_[i].guest_addr, this->areas_[i].length)) {
            std::fprintf(stderr, "bus mmio device overlaps memory area: new=[0x%lx,0x%lx) old=[0x%lx,0x%lx)\n",
                         static_cast<unsigned long>(guest_addr), static_cast<unsigned long>(end),
                         static_cast<unsigned long>(this->areas_[i].guest_addr),
                         static_cast<unsigned long>(this->areas_[i].guest_addr + this->areas_[i].length));
            return false;
        }
    }
    for (std::size_t i = 0; i < this->mmio_device_count_; i++) {
        if (ranges_overlap(guest_addr, length, this->mmio_regions_[i].guest_addr, this->mmio_regions_[i].length)) {
            std::fprintf(stderr, "bus mmio device overlaps existing device: new=[0x%lx,0x%lx)\n",
                         static_cast<unsigned long>(guest_addr), static_cast<unsigned long>(end));
            return false;
        }
    }

    device->connect_to_bus(this);
    this->mmio_regions_[this->mmio_device_count_] = {guest_addr, length, id, std::move(device)};
    this->mmio_device_count_++;
    return true;
}

bool Bus::read_mmio(std::uintptr_t guest_addr, std::size_t width, std::uint64_t *value) const {
    const MmioRegion *region = this->find_mmio_region(guest_addr, static_cast<std::uintptr_t>(width));
    if (region == nullptr || region->device == nullptr) {
        return false;
    }

    return region->device->read(guest_addr - region->guest_addr, width, value);
}

bool Bus::write_mmio(std::uintptr_t guest_addr, std::size_t width, std::uint64_t value) const {
    const MmioRegion *region = this->find_mmio_region(guest_addr, static_cast<std::uintptr_t>(width));
    if (region == nullptr || region->device == nullptr) {
        return false;
    }

    return region->device->write(guest_addr - region->guest_addr, width, value);
}

std::vector<std::uint8_t> Bus::build_dtb(const DtbConfig &config) const {
    DtbBuilder builder;
    builder.config_dtb(config);

    for (std::size_t i = 0; i < this->mmio_device_count_; i++) {
        const MmioRegion &region = this->mmio_regions_[i];
        if (region.device == nullptr) {
            continue;
        }

        region.device->config_dtb(builder, config, region.guest_addr, region.length, region.id);
    }

    return builder.finish_dtb();
}

std::uint8_t *Bus::translate(std::uintptr_t guest_addr, std::uintptr_t length) const {
    const Area *area = this->find_area(guest_addr, length);
    if (area == nullptr) {
        return nullptr;
    }

    return area->host_addr + (guest_addr - area->guest_addr);
}

const Bus::Area *Bus::find_area(std::uintptr_t guest_addr, std::uintptr_t length) const {
    std::uintptr_t end = 0;
    if (!checked_range_end(guest_addr, length, &end)) {
        return nullptr;
    }

    for (std::size_t i = 0; i < this->area_count_; i++) {
        const Area &area = this->areas_[i];
        const std::uintptr_t area_end = area.guest_addr + area.length;
        if (area.guest_addr <= guest_addr && end <= area_end) {
            return &area;
        }
    }

    return nullptr;
}

const Bus::MmioRegion *Bus::mmio_region_at(std::size_t index) const {
    if (index >= this->mmio_device_count_) {
        return nullptr;
    }

    return &this->mmio_regions_[index];
}

std::size_t Bus::mmio_device_count() const {
    return this->mmio_device_count_;
}

const Bus::MmioRegion *Bus::find_mmio_region(std::uintptr_t guest_addr, std::uintptr_t length) const {
    std::uintptr_t end = 0;
    if (!checked_range_end(guest_addr, length, &end)) {
        return nullptr;
    }

    for (std::size_t i = 0; i < this->mmio_device_count_; i++) {
        const MmioRegion &region = this->mmio_regions_[i];
        const std::uintptr_t region_end = region.guest_addr + region.length;
        if (region.guest_addr <= guest_addr && end <= region_end) {
            return &region;
        }
    }

    return nullptr;
}

void Bus::update() {
    for (std::size_t i = 0; i < this->mmio_device_count_; i++) {
        if (this->mmio_regions_[i].device != nullptr) {
            this->mmio_regions_[i].device->update();
        }
    }
}

std::size_t Bus::area_count() const {
    return this->area_count_;
}
} // namespace kvm_host

File: usertests/kvm/host-rs/src/device/bus.rs
use std::sync::{Arc, Mutex, TryLockError};
use std::time::Duration;
use tokio::task::JoinHandle;
use tokio::time;

use crate::dtb::{DtbBuilder, DtbConfig};
use crate::kvm::Kvm;

pub type BusRef = Arc<Mutex<Bus>>;
pub type DeviceRef = Arc<Mutex<dyn MmioDevice + Send>>;

pub trait MmioDevice: Send {
    fn read(&mut self, offset: usize, width: usize) -> Option<u64>;
    fn write(&mut self, offset: usize, width: usize, value: u64) -> bool;
    fn update(&mut self, bus: &Bus) {
        let _ = bus;
    }
    fn spawn_tasks(&self, _bus: BusRef, _guest_addr: usize, _length: usize, _id: u32) -> Vec<JoinHandle<()>> {
        Vec::new()
    }
    fn interrupt_pending(&self) -> bool {
        false
    }
    fn clear_interrupt(&mut self) {}
    fn config_dtb(&self, builder: &mut DtbBuilder, config: &DtbConfig, addr: usize, len: usize, id: u32);
}

#[derive(Clone)]
struct Area {
    guest_addr: usize,
    length: usize,
    host_addr: *mut u8,
}

unsafe impl Send for Area {}

#[derive(Default)]
pub struct BusBuilder {
    mmio_regions: Vec<MmioRegion>,
}

#[derive(Clone)]
pub struct MmioRegion {
    pub guest_addr: usize,
    pub length: usize,
    pub id: u32,
    pub device: DeviceRef,
}

pub struct Bus {
    areas: Vec<Area>,
    mmio_regions: Box<[MmioRegion]>,
}

pub fn checked_range_end(start: usize, length: usize) -> Option<usize> {
    (length != 0).then(|| start.checked_add(length)).flatten()
}

fn ranges_overlap(left_start: usize, left_length: usize, right_start: usize, right_length: usize) -> bool {
    match (
        checked_range_end(left_start, left_length),
        checked_range_end(right_start, right_length),
    ) {
        (Some(left_end), Some(right_end)) => left_start < right_end && right_start < left_end,
        _ => true,
    }
}

impl BusBuilder {
    pub fn add_mmio_device(
        &mut self,
        guest_addr: usize,
        length: usize,
        device: DeviceRef,
        id: u32,
    ) -> Result<(), String> {
        let Some(end) = checked_range_end(guest_addr, length) else {
            return Err(format!(
                "bus mmio device has invalid range: addr=0x{guest_addr:x} length=0x{length:x}"
            ));
        };
        if self
            .mmio_regions
            .iter()
            .any(|region| ranges_overlap(guest_addr, length, region.guest_addr, region.length))
        {
            return Err(format!(
                "bus mmio device overlaps existing device: new=[0x{guest_addr:x},0x{end:x})"
            ));
        }
        self.mmio_regions.push(MmioRegion {
            guest_addr,
            length,
            id,
            device,
        });
        Ok(())
    }

    pub fn build(self) -> Bus {
        Bus {
            areas: Vec::new(),
            mmio_regions: self.mmio_regions.into_boxed_slice(),
        }
    }
}

impl Bus {
    pub fn mmio_regions(&self) -> &[MmioRegion] {
        &self.mmio_regions
    }

    pub fn map_area(&mut self, kvm: &Kvm, guest_addr: usize, length: usize) -> Result<*mut u8, String> {
        let Some(end) = checked_range_end(guest_addr, length) else {
            return Err(format!(
                "bus map area has invalid range: addr=0x{guest_addr:x} length=0x{length:x}"
            ));
        };
        if self
            .areas
            .iter()
            .any(|area| ranges_overlap(guest_addr, length, area.guest_addr, area.length))
        {
            return Err(format!(
                "bus map area overlaps existing range: new=[0x{guest_addr:x},0x{end:x})"
            ));
        }
        if self
            .mmio_regions
            .iter()
            .any(|region| ranges_overlap(guest_addr, length, region.guest_addr, region.length))
        {
            return Err(format!(
                "bus map area overlaps mmio device: addr=0x{guest_addr:x} length=0x{length:x}"
            ));
        }
        let host_addr = kvm.map_area_raw(guest_addr, length)?;
        self.areas.push(Area {
            guest_addr,
            length,
            host_addr,
        });
        Ok(host_addr)
    }

    pub fn translate(&self, guest_addr: usize, length: usize) -> Option<*mut u8> {
        let end = checked_range_end(guest_addr, length)?;
        self.areas.iter().find_map(|area| {
            let area_end = area.guest_addr.checked_add(area.length)?;
            if area.guest_addr <= guest_addr && end <= area_end {
                Some(unsafe { area.host_addr.add(guest_addr - area.guest_addr) })
            } else {
                None
            }
        })
    }

    pub fn read_mmio(&self, guest_addr: usize, width: usize) -> Option<u64> {
        let region = self.find_mmio_region(guest_addr, width)?;
        region
            .device
            .lock()
            .expect("mmio device lock poisoned")
            .read(guest_addr - region.guest_addr, width)
    }

    pub fn write_mmio(&self, guest_addr: usize, width: usize, value: u64) -> bool {
        let Some(region) = self.find_mmio_region(guest_addr, width) else {
            return false;
        };
        region
            .device
            .lock()
            .expect("mmio device lock poisoned")
            .write(guest_addr - region.guest_addr, width, value)
    }

    fn find_mmio_region(&self, guest_addr: usize, length: usize) -> Option<&MmioRegion> {
        let end = checked_range_end(guest_addr, length)?;
        self.mmio_regions.iter().find(|region| {
            let region_end = region.guest_addr + region.length;
            region.guest_addr <= guest_addr && end <= region_end
        })
    }

    pub fn external_interrupt_pending(&self) -> bool {
        self.mmio_regions.iter().any(|region| {
            region.id == 0
                && region
                    .device
                    .lock()
                    .expect("mmio device lock poisoned")
                    .interrupt_pending()
        })
    }

    pub fn update(&self) {
        for region in &self.mmio_regions {
            region.device.lock().expect("mmio device lock poisoned").update(self);
        }
    }

    pub fn spawn_runtime_tasks(bus: BusRef) -> Result<RuntimeTasks, String> {
        let regions: Vec<(usize, usize, u32, DeviceRef)> = {
            let bus = bus.lock().map_err(|_| "kvm bus lock poisoned".to_string())?;
            bus.mmio_regions
                .iter()
                .map(|region| (region.guest_addr, region.length, region.id, region.device.clone()))
                .collect()
        };

        let mut handles = Vec::with_capacity(regions.len());
        for (guest_addr, length, id, device) in regions {
            handles.extend(
                device
                    .lock()
                    .map_err(|_| "mmio device lock poisoned".to_string())?
                    .spawn_tasks(bus.clone(), guest_addr, length, id),
            );
        }
        Ok(RuntimeTasks { handles })
    }

    pub async fn notify(bus: &BusRef) -> Result<(), String> {
        loop {
            if Self::try_notify(bus)? {
                return Ok(());
            }
            time::sleep(Duration::from_millis(1)).await;
        }
    }

    fn try_notify(bus: &BusRef) -> Result<bool, String> {
        match bus.try_lock() {
            Ok(bus) => {
                bus.update();
                Ok(true)
            }
            Err(TryLockError::WouldBlock) => Ok(false),
            Err(TryLockError::Poisoned(_)) => Err("kvm bus lock poisoned".to_string()),
        }
    }

    pub fn build_dtb(&self, config: &DtbConfig) -> Vec<u8> {
        let mut builder = DtbBuilder::default();
        builder.config_dtb(config);
        for region in &self.mmio_regions {
            region.device.lock().expect("mmio device lock poisoned").config_dtb(
                &mut builder,
                config,
                region.guest_addr,
                region.length,
                region.id,
            );
        }
        builder.finish_dtb()
    }
}

pub struct RuntimeTasks {
    handles: Vec<JoinHandle<()>>,
}

impl Drop for RuntimeTasks {
    fn drop(&mut self) {
        for handle in &self.handles {
            handle.abort();
        }
    }
}
